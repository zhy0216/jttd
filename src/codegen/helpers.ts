import ts from "typescript";
import { Schema, Type } from "~/types.ts";
import { LiteralEmitter } from "~/codegen/literal.ts";

export const getSchemaByTypeNode = (
  typeNode: ts.TypeNode,
  option: { checker: ts.TypeChecker },
): Schema => {
  const typeKind = typeNode.kind;
  if (typeKind === ts.SyntaxKind.NumberKeyword) {
    return { type: Type.number };
  }

  if (typeKind === ts.SyntaxKind.UndefinedKeyword) {
    return { type: Type.undefined };
  }

  if (typeKind === ts.SyntaxKind.BooleanKeyword) {
    return { type: Type.boolean };
  }

  if (typeKind === ts.SyntaxKind.LiteralType) {
    const schema = new LiteralEmitter({
      node: typeNode as ts.LiteralTypeNode,
      checker: option.checker,
    }).emit();

    return { ...schema };
  }

  if (typeKind === ts.SyntaxKind.ArrayType) {
    const schema = getSchemaByTypeNode(
      (typeNode as ts.ArrayTypeNode).elementType,
      option,
    );

    return { type: Type.array, elementType: schema };
  }

  throw new Error(`wrong Node: ${typeKind}`);
};
