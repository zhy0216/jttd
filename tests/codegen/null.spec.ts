import { test, expect } from "bun:test";
import { genFromFixture } from "tests/helpers.ts";

test("null.ts", () => {
  expect(genFromFixture("null.ts")).toMatchSnapshot();
  expect(genFromFixture("null.ts")).toMatchSnapshot();
});

test("null222.ts", () => {
  expect(genFromFixture("null.ts")).toMatchSnapshot();
  expect(genFromFixture("null.ts")).toMatchSnapshot();
});
