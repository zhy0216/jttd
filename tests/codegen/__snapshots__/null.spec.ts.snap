// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`null.ts 1`] = `undefined`;

exports[`null.ts 2`] = `undefined`;

exports[`null222.ts 1`] = `undefined`;

exports[`null222.ts 2`] = `undefined`;
